control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs

```diff
@@ -130,57 +139,282 @@ async fn nexus_reconciler(
+    nexus: &mut OperationGuardArc<NexusSpec>,
+    context: &PollContext,
+) -> PollResult {
+    let rebuild_state = nexus.as_ref().rebuild_state.clone();
+    for (uri, state) in rebuild_state.iter() {
+        if state.stage == RebuildStage::PartialRebuild {
+            let nexus_uuid = nexus.uuid();
+            let nexus_state = context.registry().nexus(nexus_uuid).await?;
+            for ch in nexus_state.children.iter() {
+                if &ch.uri == uri {
+                    match (ch.state.clone(), ch.state_reason.clone()) {
+                        (ChildState::Online, _) => {
+                            info!(
+                                "Partial rebuild completed for child: {:?}, removing rebuild info",
+                                uri
+                            );
+                            let _ = nexus
+                                .set_rebuild_state(context.registry(), uri.clone(), None)
+                                .await;
+                        }
+                        (
+                            ChildState::Degraded | ChildState::Unknown,
+                            ChildStateReason::OutOfSync,
+                        ) => {
+                            info!("Partial rebuild in progress for child :{:?}", uri);
+                            break;
+                        }
+                        (_, _) => {
+                            info!("Encountered failure during Partial rebuild, Child: {:?} Faulted again. Initiating Full rebuild for the child", uri);
+                            nexus
+                                .set_rebuild_state(
+                                    context.registry(),
+                                    uri.clone(),
+                                    Some(RebuildInfo::new(None, RebuildStage::FullRebuildInit)),
+                                )
+                                .await?;
+                        }
+                    }
+                }
+            }
+        }
+    }
+    Ok(PollerState::Idle)
+}
+
+/// Iterates over child rebuild stages. If stage is FullRebuildInit, Removes child from the Nexus
+/// children list and removes rebuild stage for the child from hashmap.
+#[tracing::instrument(skip(nexus, context), level = "trace", fields(nexus.uuid = %nexus.uuid(), request.reconcile = true))]
+pub(super) async fn faulted_children_remover(
+    nexus: &mut OperationGuardArc<NexusSpec>,
+    context: &PollContext,
+) -> PollResult {
+    let nexus_uuid = nexus.uuid();
+    let nexus_state = context.registry().nexus(nexus_uuid).await?;
+    let child_count = nexus_state.children.len();
+    if nexus_state.status == NexusStatus::Degraded && child_count > 1 {
+        let span = tracing::info_span!("faulted_children_remover", nexus.uuid = %nexus_uuid, request.reconcile = true);
+        async {
+            let nexus_spec_clone = nexus.lock().clone();
+            for (child, state) in nexus_spec_clone.rebuild_state.iter() {
```

Comment:
Instead of iterating the rebuild_state hashmap, let it iterate over nexus_state.children as it does today. Then for each children, get  RebuildInfo from hashmap by child key.

Reply:
That will increase number of operations right? By iterating on stages we check if we need to proceed further on the reconciler by checking stage. In this case we more further if stage == FullRebuildInit meaning remove child from nexus. This way we wont do any operations that is not required for a child in stage != FullRebuildInit

If this reconciler runs on healthy nexus which will have empty hashmap. We will exit early.

Reply:
`That will increase number of operations right`
We're only dealing with 3-4 child so doesn't matter. Moreover, we'd be comparing against state as it is today instead of spec.

Reply:
Removed stages from implementation.